feat(avm)!: ALU Shifts

MirandaWood opened the pull request:
## a << b OR a >> b

Adds shift opcodes `SHL` and `SHR` following some of the logic [here](https://github.com/AztecProtocol/aztec-packages/blame/bf7a99d8d8e3faa00c9395a17acf7f92b171a7a1/barretenberg/cpp/pil/avm/alu.pil#L518).

We reuse the decomposition columns for u128 mul & div to show:

- `SHL`: show that `a = 2^(max_bits - shift) * a_hi + a_lo` and assign `c = a_lo * 2^shift`
  - includes range checking `a_hi` is less than `shift` bits and `a_lo` less than `max_bits - shift` bits
- `SHR`: show that `a = 2^shift * a_hi + a_lo` and assign `c = a_hi`
   - includes range checking `a_hi` is less than `max_bits - shift` bits and `a_lo` less than `shift` bits

### Bit Size Overflow

If `shift > max_bits`, then `max_bits - shift` (used in both cases in the range checks) will underflow the field and fail. Since we just need to return `c = 0` in this case, I set `sel_shift_ops_no_overflow` as a gate for assigning `c`. If we have an overflow, we assign `a_lo = shift - max_bits` and range check it against `max_bits`. The range check proves we have a shift overflow (i.e. `SHIFT_OVERFLOW = sel_shift_ops * (1 - sel_shift_ops_no_overflow) <==> shift > max_bits`) and the result is assigned as 0. e.g. for `SHR`:

```
ic - sel_shift_ops_no_overflow * a_hi = 0;
```

### Errors

Like `DIV`, shift operations on `FF` values are invalid, so I reused columns/relations to assign `FF_TAG_ERR = 1` in this case. Mismatched `a` and `b` tags (to follow new noir style) also give an error (`sel_ab_tag_mismatch`). As before:

```
sel_tag_err = sel_ab_tag_mismatch + FF_TAG_ERR - sel_ab_tag_mismatch * FF_TAG_ERR;
```

### Notes

 - To avoid more lookups than necessary, I use `precomputed.power_of_2` (to find the limb size of `a_lo`, either `2^shift` or `2^(max_bits - shift)`) once, then for `SHL` when we also need `2^shift`, I use `helper1` to inject the value and prove correctness in a relation. It might be cheaper in another way, or just clearer to always lookup the `2^shift` value and assign `helper1 = 2^(max_bits - shift)` - happy to change/discuss!
 - Made some changes to `TaggedValue` behaviour to fix issues with overflowing the bit size - thank you Facundo for your help on this!

IlyasRidhuan commented on barretenberg/cpp/pil/vm2/alu.pil:
```diff
@@ -150,6 +150,52 @@ CHECK_AB_TAGS * ( (ia_tag - ib_tag) * ( AB_TAGS_EQ * (1 - ab_tags_diff_inv) + ab
+pol TWO_POW_64 = 2 ** 64;
+
+// Reusing columns for decomposition (#[OP_ID_CHECK] ensures selectors are mutually exclusive):
+pol DECOMPOSED_A = ((sel_mul_u128 + sel_shift_ops_no_overflow) * ia) + (sel_shift_ops - sel_shift_ops_no_overflow) * (ib - max_bits) + (sel_is_u128 * sel_op_div * (1 - sel_tag_err) * ic);
```
I don't see the relation where `sel_shift_ops_no_overflow = 1` IFF this is a shift op? Do we need one?

MirandaWood replied:
Oh whoops, yes we probably do. I forgot about this after I started using `pol SHIFT_OVERFLOW = sel_shift_ops * (1 - sel_shift_ops_no_overflow);` (to avoid two columns). Would it be better to add a relation or revert `SHIFT_OVERFLOW` to a column?

IlyasRidhuan replied:
i think you can probably do
`sel_shift_ops_no_overflow * (1 - sel_shift_ops) = 0`?

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/simulation/alu.cpp:
```diff
@@ -177,6 +177,56 @@ MemoryValue Alu::op_not(const MemoryValue& a)
+    } catch (const TagMismatchException& e) {
+        events.emit({ .operation = AluOperation::SHR, .a = a, .b = b, .error = AluError::TAG_ERROR });
+        throw AluException("SHR, " + std::string(e.what()));
+    } catch (const InvalidOperationTag& e) {
```
im surprised we don't do anything different for when theres a tag mismatch or theres an invalid tag tbh. But if this is correct, these two catches can probably just be collapsed into 1 catch

MirandaWood replied:
I did try this but couldn't find a nice c++ way of doing it, any suggestions?
I was thinking we should separate `AluError::TAG_ERROR` into at least tag mismatches and invalid ops, but wasn't sure it was worth the effort considering the circuit does pretty much the same in both cases (`sel_tag_err = 1` -> match with `execution.sel_err`)

IlyasRidhuan replied:
if the circuit doesnt do anything different then i think it's fine to just collapse them into 1

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/simulation/alu.cpp:
```diff
@@ -177,6 +177,56 @@ MemoryValue Alu::op_not(const MemoryValue& a)
+        range_check.assert_range(a_lo, a_lo_bits);
+        range_check.assert_range(static_cast<uint128_t>(a.as_ff()) >> a_lo_bits,
+                                 overflow ? tag_bits : static_cast<uint8_t>(b_num));
+        events.emit({ .operation = AluOperation::SHL, .a = a, .b = b, .c = c });
```
it's sad there's so much circuit leakage in here...i guess the range checks are unavoidable but perhaps long term we might want to look into some helpers for decompositions

MirandaWood replied:
I'd be happy to add a helper like the ones used for fixed size decomps e.g.
```
            // For u128, we decompose c and b into 64 bit chunks and discard the highest bits given by the product:
            auto c_decomp = decompose(static_cast<uint128_t>(c.as_ff()));
            auto b_decomp = decompose(static_cast<uint128_t>(b.as_ff()));
            range_check.assert_range(c_decomp.lo, 64)....
```
The finicky part is when we overflow the bit size, since I want to reuse the range check I need to set a different `a_lo`, so it's unavoidable here to leak that case :(

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/simulation/execution.cpp:
```diff
@@ -212,6 +212,50 @@ void Execution::op_not(ContextInterface& context, MemoryAddress src_addr, Memory
+        throw OpcodeExecutionException("SHL Exception: " + std::string(e.what()));
+    } catch (const std::exception& e) {
+        // We have some err not handled by TAG_ERROR:
+        throw OpcodeExecutionException("Alu SHL operation failed");
```
Err im not sure this is correct. The `OpcodeExecutionException` isnt catastrophic so the execution will proceed. This will create an entry in the execution trace without a corresponding alu entry. This will almost certainly break the lookup.

IlyasRidhuan replied:
If you want to keep this catch, you should just re-throw a runtime error i think. That way it will fall into the `catch (const std::exception& e)` in `Execution::execute`

Commit:
feat(avm)!: WIP ALU Shifts

## a << b OR a >> b

Adds shift opcodes `SHL` and `SHR` following some of the logic [here](https://github.com/AztecProtocol/aztec-packages/blame/bf7a99d8d8e3faa00c9395a17acf7f92b171a7a1/barretenberg/cpp/pil/avm/alu.pil#L518).

We reuse the decomposition columns for u128 mul & div to show:

- `SHL`: show that `a = 2^(max_bits - shift) * a_hi + a_lo` and assign `c = a_lo * 2^shift`
  - includes range checking `a_hi` is less than `shift` bits and `a_lo` less than `max_bits - shift` bits
- `SHR`: show that `a = 2^shift * a_hi + a_lo` and assign `c = a_hi`
   - includes range checking `a_hi` is less than `max_bits - shift` bits and `a_lo` less than `shift` bits

### Bit Size Overflow

If `shift > max_bits`, then `max_bits - shift` (used in both cases in the range checks) will underflow the field and fail. Since we just need to return `c = 0` in this case, I set `sel_shift_ops_no_overflow` as a gate for assigning `c`. If we have an overflow, we assign `a_lo = shift - max_bits` and range check it against `max_bits`. The range check proves we have a shift overflow (i.e. `SHIFT_OVERFLOW = sel_shift_ops * (1 - sel_shift_ops_no_overflow) <==> shift > max_bits`) and the result is assigned as 0. e.g. for `SHR`:

```
ic - sel_shift_ops_no_overflow * a_hi = 0;
```

### Errors

Like `DIV`, shift operations on `FF` values are invalid, so I reused columns/relations to assign `FF_TAG_ERR = 1` in this case. Mismatched `a` and `b` tags (to follow new noir style) also give an error (`sel_ab_tag_mismatch`). As before:

```
sel_tag_err = sel_ab_tag_mismatch + FF_TAG_ERR - sel_ab_tag_mismatch * FF_TAG_ERR;
```

### Notes

 - To avoid more lookups than necessary, I use `precomputed.power_of_2` (to find the limb size of `a_lo`, either `2^shift` or `2^(max_bits - shift)`) once, then for `SHL` when we also need `2^shift`, I use `helper1` to inject the value and prove correctness in a relation. It might be cheaper in another way, or just clearer to always lookup the `2^shift` value and assign `helper1 = 2^(max_bits - shift)` - happy to change/discuss!
 - Made some changes to `TaggedValue` behaviour to fix issues with overflowing the bit size - thank you Facundo for your help on this!

Co-authored-by: MirandaWood <miranda@aztecprotocol.com>
